### PR TITLE
adds `allElements()` method in Value API

### PR DESCRIPTION
**Pull request by desaikd**
*Issue #693*

*Description of changes:*
This PR works on adding `allElements()` along with `elements()` for Value API.

*Test:*
Adds unit tests for `allElements()`.

### CODE REVIEW

File: src/dom/Struct.ts [hunk 1]
```diff
@@ -141,12 +141,16 @@ export class Struct extends Value(
...
     let singleValueFields = Object.create(null);
-    for (const [fieldName, values] of this.allFields()) {
-      singleValueFields[fieldName] = values[values.length - 1];
+    for (const [fieldName, value] of this.fields()) {
+      singleValueFields[fieldName] = value;
     }
     return Object.values(singleValueFields);
   }
 
+  allElements(): Value[][] {
```
Comment (zslayton) [inline src/dom/Struct.ts:150]: This is returning an array-of-arrays, but I think we want it to have the same method signature as `elements` on line 142 above.

You can use the [`Array` method `flat()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) to achieve this.
```ts
return Object.values(this._fields).flat();
```

File: src/dom/Value.ts [hunk 2]
```diff
@@ -159,9 +159,32 @@ export interface Value {
...
 
+  /**
+   * For the Struct, List, and SExpression types, returns an array containing the container's
+   * nested values; otherwise throws an Error.
+   * If there are duplicate fields in a Struct, then it only returns all the values for that field.
+   *
+   * Like the `elements()` method, but returns an array containing the field name/value pairs with *all* of
+   * the associated values to a field name, instead of field name/value pairs with just the last associated value
+   * to a field name. For example, for the
+   * following struct:
```
Comment (zslayton) [inline src/dom/Value.ts:174]: Ah, now I understand why `Struct#allElements` was returning `Value[][]`.

The `Value#elements` method makes the most sense when you're working with a `list` or `sexp`. In that use case, it's straightforward: it gives you all of the child elements of the sequence.

For `struct`s, it gets weird. It only returns the field values. If a user wanted the name/value(s) pairs, they could use `fields()` or `allFields()` to get them. Arguably, we could've made `elements()` just throw an exception if you called it on a `struct`, but we decided to make it act more or less like the other containers.

For consistency with its behavior for `list` and `sexp`, I think this should return a flat `Value[]` with a list of the struct's field values.

File: src/dom/Value.ts [hunk 3]
```diff
@@ -159,9 +159,32 @@ export interface Value {
   /**
    * For the Struct, List, and SExpression types, returns an array containing the container's
    * nested values; otherwise throws an Error.
+   * If there are duplicate fields in a Struct, then it only returns the last value for that field.
    */
   elements(): Value[];
 
+  /**
+   * For the Struct, List, and SExpression types, returns an array containing the container's
```
Comment (zslayton) [inline src/dom/Value.ts:167]: This PR doesn't seem to implement `allElements()` for `Struct` and `List`. Am I missing it?

Comment (desaikd) [reply src/dom/Value.ts:167]: That should be just `Struct` not `SExpression` and `List`. I will change it.

**Comment by desaikd** [comment]
Closing this with the intention to express `allElements()` behavior in `elements()` with `v5.0.0` release.

**Pull request closed without merging**